core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java

```diff
@@ -0,0 +1,369 @@
+      establishConnectionToLeader();
+    });
+  }
+
+  @Override
+  public void stop() throws IOException {
+    LOG.info("Stopping backup worker role.");
+    // Close leader close listener.
+    // This will ensure, connection won't be re-established when closed during stop.
+    if (mLeaderConnectionCloseListener != null) {
+      mLeaderConnectionCloseListener.close();
+    }
+    // Close the connection with the leader.
+    if (mLeaderConnection != null) {
+      mLeaderConnection.close();
+      mLeaderConnection = null;
+    }
+    // Cancel ongoing backup task.
+    if (mBackupFuture != null) {
+      mBackupFuture.cancel(true);
+    }
+    // Cancel heartbeat task.
+    if (mBackupProgressFuture != null) {
+      mBackupProgressFuture.cancel(true);
+    }
+    // Cancel suspend timeout.
+    if (mSuspendTimeoutFuture != null && !mSuspendTimeoutFuture.isDone()) {
+      mSuspendTimeoutFuture.cancel(true);
+    }
+    // Stopping the base after because closing connection uses the base executor.
+    super.stop();
+  }
+
+  @Override
+  public Map<ServiceType, GrpcService> getRoleServices() {
+    return Collections.emptyMap();
+  }
+
+  @Override
+  public BackupStatus backup(BackupPRequest request) throws AlluxioException {
+    throw new IllegalStateException("Worker role can't serve RPCs");
+  }
+
+  @Override
+  public BackupStatus getBackupStatus() throws AlluxioException {
+    throw new IllegalStateException("Worker role can't serve RPCs");
+  }
+
+  /**
+   * Handler for suspend message. It's used in secondary master.
+   */
+  private void handleSuspendJournalsMessage(BackupSuspendMessage suspendMsg) {
+    LOG.info("Received suspend message: {}", suspendMsg.toString());
+    Preconditions.checkState(mSuspendTimeoutFuture == null || mSuspendTimeoutFuture.isDone(),
+        "Already suspended.");
+    try {
+      mJournalSystem.suspend();
+      LOG.info("Suspended journals for backup.");
+    } catch (IOException e) {
+      String failMessage = "Failed to suspended journals for backup.";
+      LOG.error(failMessage, e);
+      throw new RuntimeException(failMessage, e);
+    }
+    // Schedule a task to resume journals if protocol is not followed by the leader.
+    mSuspendTimeoutFuture = mTaskScheduler.schedule(() -> {
+      try {
+        LOG.info("Resuming journals as backup request hasn't been received.");
+        mJournalSystem.resume();
+      } catch (IOException e) {
+        ProcessUtils.fatalError(LOG, e, "Failed to resume journals.");
+      }
+    }, BACKUP_ABORT_AFTER_SUSPEND_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Handler for backup request message. It's used in secondary master.
+   */
+  private void handleRequestMessage(BackupRequestMessage requestMsg) {
+    LOG.info("Received backup message: {}", requestMsg.toString());
+    Preconditions.checkState(!mBackupTracker.inProgress(), "Backup in progress");
+
+    // Cancel timeout timer.
+    mSuspendTimeoutFuture.cancel(true);
```

Comment:
Is it possible that this worker failed to receive the suspend message but receive the request message?
Should we add a preconditions.checkState or add a null check?

Reply:
Backup-leader don't send request before sending suspend message.